src/base/packable_object.hpp

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (C) 2008 Search Solution Corporation. All rights reserved by Search Solution.
+ *
+ *   This program is free software; you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation; either version 2 of the License, or
+ *   (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ */
+
+/*
+ * packable_object.hpp
+ */
+
+#ifndef _PACKABLE_OBJECT_HPP_
+#define _PACKABLE_OBJECT_HPP_
+
+#ident "$Id$"
+
+#include "packer.hpp"
+#include <map>
+
+namespace cubpacking
+{
+
+  class packer;
+
+  class packable_object
+  {
+    public:
+      virtual int pack (packer *serializator) = 0;
+      virtual int unpack (packer *serializator) = 0;
+
+      virtual bool is_equal (const packable_object *other) = 0;
+
+      /* used at packing to get info on how much memory to reserve */
+      virtual size_t get_packed_size (packer *serializator) = 0;
+  };
+
+  class self_creating_object
```

Comment:
We discussed self_creating_object/object_builder. We concluded that it is necessary to pack/unpack virtual objects.

Your initial idea was to pack/unpack multiple pointers inside an object with object builder. So we have two options:

  1. To have specialized builders for each type of virtual pointers.
  2. To have builders capable of handling multiple types of object.

In both cases, I think this should be separated of packable_object.hpp. I propose separating this topic to a different pull request.